pkg/client/unversioned/discovery_client.go

```diff
@@ -0,0 +1,172 @@
+	// preferred version.
+	ServerGroups() (*unversioned.APIGroupList, error)
+}
+
+// ServerResourcesInterface has methods for obtaining supported resources on the API server
+type ServerResourcesInterface interface {
+	// ServerResourcesForGroupVersion returns the supported resources for a group and version.
+	ServerResourcesForGroupVersion(groupVersion string) (*unversioned.APIResourceList, error)
+	// ServerResources returns the supported resources for all groups and versions.
+	ServerResources() (map[string]*unversioned.APIResourceList, error)
+}
+
+// DiscoveryClient implements the functions that dicovery server-supported API groups,
+// versions and resources.
+type DiscoveryClient struct {
+	httpClient HTTPClient
+	baseURL    url.URL
+}
+
+// Convert unversioned.APIVersions to unversioned.APIGroup. APIVersions is used by legacy v1, so
+// group would be "".
+func apiVersionsToAPIGroup(apiVersions *unversioned.APIVersions) (apiGroup unversioned.APIGroup) {
+	groupVersions := []unversioned.GroupVersion{}
+	for _, version := range apiVersions.Versions {
+		groupVersion := unversioned.GroupVersion{
+			GroupVersion: version,
+			Version:      version,
+		}
+		groupVersions = append(groupVersions, groupVersion)
+	}
+	apiGroup.Versions = groupVersions
+	// There should be only one groupVersion returned at /api
+	apiGroup.PreferredVersion = groupVersions[0]
+	return
+}
+
+func (d *DiscoveryClient) get(url string) (resp *http.Response, err error) {
+	req, err := http.NewRequest("GET", url, nil)
+	if err != nil {
+		return nil, err
+	}
+	return d.httpClient.Do(req)
+}
+
+// ServerGroups returns the supported groups, with information like supported versions and the
+// preferred version.
+func (d *DiscoveryClient) ServerGroups() (apiGroupList *unversioned.APIGroupList, err error) {
+	// Get the groupVersions exposed at /api
+	resp, err := d.get(d.baseURL.String() + "/api")
+	if err != nil {
+		return nil, err
+	}
+	var v unversioned.APIVersions
+	defer resp.Body.Close()
+	err = json.NewDecoder(resp.Body).Decode(&v)
+	if err != nil {
+		return nil, fmt.Errorf("unexpected error: %v", err)
+	}
+	apiGroup := apiVersionsToAPIGroup(&v)
+
+	// Get the groupVersions exposed at /apis
+	resp2, err := d.get(d.baseURL.String() + "/apis")
+	if err != nil {
+		return nil, err
+	}
+	defer resp2.Body.Close()
+	apiGroupList = &unversioned.APIGroupList{}
+	if err = json.NewDecoder(resp2.Body).Decode(&apiGroupList); err != nil {
+		return nil, fmt.Errorf("unexpected error: %v", err)
+	}
+
+	// append the group retrieved from /api to the list
+	apiGroupList.Groups = append(apiGroupList.Groups, apiGroup)
+	return apiGroupList, nil
+}
+
+// ServerResroucesForGroupVersion returns the supported resources for a group and version.
+func (d *DiscoveryClient) ServerResourcesForGroupVersion(groupVersion string) (resources *unversioned.APIResourceList, err error) {
+	var url string
+	if groupVersion == "v1" {
+		url = d.baseURL.String() + "/api/" + groupVersion
```

Comment:
In general--please don't build URLs with concatenation, make a url.URL and
put the appropriate things in the appropriate places... :)

On Fri, Oct 16, 2015 at 1:53 PM, krousey <notifications@github.com> wrote:

> In pkg/client/unversioned/discovery_client.go
> <https://github.com/kubernetes/kubernetes/pull/15659#discussion_r42288992>
> :
>
> > +	defer resp2.Body.Close()
> > +	apiGroupList = &unversioned.APIGroupList{}
> > +	if err = json.NewDecoder(resp2.Body).Decode(&apiGroupList); err != nil {
> > +		return nil, fmt.Errorf("unexpected error: %v", err)
> > +	}
> > +
> > +	// append the group retrieved from /api to the list
> > +	apiGroupList.Groups = append(apiGroupList.Groups, apiGroup)
> > +	return apiGroupList, nil
> > +}
> > +
> > +// ServerResroucesForGroupVersion returns the supported resources for a group and version.
> > +func (d *DiscoveryClient) ServerResourcesForGroupVersion(groupVersion string) (resources *unversioned.APIResourceList, err error) {
> > +	var url string
> > +	if groupVersion == "v1" {
> > +		url = d.baseURL.String() + "/api/" + groupVersion
>
> Is is possible that baseURL has a non-empty RawQuery or RawFragment? Is
> so, this string concatenation won't work.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/15659/files#r42288992>.
>